### PR TITLE
Adding traffic shaping support for CNI network driver

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

Adding traffic shaping support for CNI network driver - it's also a sub-task of kubenet deprecation work.

Design document is available here: https://github.com/kubernetes/community/pull/1893

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

/cc @freehan @jingax10 @caseydavenport @dcbw 

/sig network
/sig node

**Release note**:

```release-note
Support traffic shaping for CNI network driver
```

### CODE REVIEW

**Comment by thockin** [comment]
This looks OK, but I'll need someone to really try it out.  What can we do about testing?

**Comment by m1093782566** [comment]
/assign @rramkumar1 

Can you take a look when you have chance? :)

Thanks!

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (dcbw) [inline pkg/kubelet/dockershim/network/cni/cni.go]: Should probably also mention:
https://github.com/containernetworking/plugins/blob/master/plugins/meta/bandwidth/README.md

**Review by dcbw — APPROVED** [review]

**Comment by Lion-Wei** [comment]
@dcbw ,I think we better have some e2e test case about `bandwidth`, but cni plugin v0.8 have not been released yet, so how to get `bandwidth` package in our test, that'a a problem. : (

**Comment by m1093782566** [comment]
As @Lion-Wei mentioned, my team have been working on adding e2e tests for some days. We find

* There is no specific job for running CNI-plugins-related test cases. We need to create new test job in kubernetes/test-infra.

* If we reuse the test jobs `kubeadm-gce-cni-calico` and `kubeadm-gce-cni-flannel`, we should update their `--ginkgo.focus`. We need to update the test job in kubernetes/test-infra as well.

So, can we let this PR in first and then update the jobs in kubernetes/test-infra. My team will continue to work on it.

BTW, the functionalities of cni bandwidth plugins is already being tested in `containernetworking/plugins` repo, so can we just write some UTs in kubernetes repo for now?

**Comment by m1093782566** [comment]
@thockin @dcbw 

I just added some tests, PTAL.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: This was a bug in the CNI documentation; it is actually bits per second.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: This value should be hard-coded to 1600 bits - it's the linux TC default and is what kubenet uses.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: The documentation was wrong; the units are bits per second.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: bits per second here

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: should be the default of 1600

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (squeed) [inline pkg/kubelet/dockershim/network/cni/cni.go]: All of these fields should have `omitEmpty`

**Review by squeed — CHANGES_REQUESTED** [review]

**Comment by squeed** [comment]
I did a manual end-to-end test as part of #64445 and found some bugs - left some comments.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Fixed. Thanks!

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done. Thanks!

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done.

**Comment by m1093782566** [comment]
@squeed 

Thanks for your comments, we have addressed all of them. PTAL.

BTW, I found there are some bandwidth-related changes in #64445 as well. As my friend and I started this work 2 months ago and spend lots of time, would you please let this PR in first and rework #64445?

**Review by squeed — APPROVED** [review]

**Comment by m1093782566** [comment]
@squeed 

Thank you for your consideration :)

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Done.

**Comment by squeed** [comment]
@kubernetes/sig-network-pr-reviews this is ready for review

**Comment by m1093782566** [comment]
ping @thockin @freehan for approval

**Comment by m1093782566** [comment]
/retest

/assign @thockin back for final review - I already got a lgtm from @dcbw :)

PS.  PR #64445 is still waiting for this PR getting in first, so would you prioritize it?

**Comment by m1093782566** [comment]
All CI  checks are green. I think this PR is ready to go.

**Comment by feiskyer** [comment]
lgtm
/retest

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (thockin) [inline pkg/kubelet/dockershim/network/cni/cni.go]: `in b` is ambiguous.  Please spell out "bits" or "bytes"

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (thockin) [inline pkg/kubelet/dockershim/network/cni/cni.go]: `in bps` is ambiguous.  Please spell out "bits" or "bytes"

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (thockin) [inline pkg/kubelet/dockershim/network/cni/cni.go]: so it is required but not used?  Can't you just assume a default value of 0?

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: bits here. Done

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: bits per second. Done.

File: pkg/kubelet/dockershim/network/cni/cni.go
Comment (m1093782566) [reply pkg/kubelet/dockershim/network/cni/cni.go]: > so it is required but not used?

Yes.

> Can't you just assume a default value of 0?

Okay.

**Comment by m1093782566** [comment]
@thockin

 I just pushed an update. PTAL.

**Comment by m1093782566** [comment]
Thanks @Lion-Wei 

/hold

for @thockin final review :)

**Comment by timchenxiaoyu** [comment]
does support change bandwith  dynamic when pods  in runtime ?

**Comment by m1093782566** [comment]
Currently, it doesn't support - we configure bandwidth limit in pod start up. Of course, we can support it if we get better consensus.

**Comment by m1093782566** [comment]
THANKS @thockin

**Pull request merged**
